flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -425,6 +462,13 @@ public void disconnectJobManager(final JobID jobId, final Exception cause) {
 					slotRequest.getJobId(),
 					slotRequest.getAllocationId());
 
+				if (failureRater.exceedMaximumFailureRate()) {
+					return FutureUtils.completedExceptionally(new MaximumFailedTaskManagerExceedingException(
+						new RuntimeException(String.format("Maximum number of failed container %d in interval %s "
+							+ " is detected in Resource Manager.", failureRater.getCurrentFailureRate(),
+							failureRater.getFailureInterval().toString()))));
+				}
```

Comment:
I think this is the wrong place to check this. We should check the `failureRater.exceedMaximumFailureRater()` in a method from which we call `startNewWorker`. E.g. `tryStartNewWorker`.

```
protected Collection<ResourceProfile> tryStartNewWorker(ResourceProfile resourceProfile) {
	if (failureRater.exceedMaximumFailureRate()) {
		return Collections.emptyList();
	}

	return startNewWorker(resourceProfile);
}

protected void startNewWorkerIfNeeded(ResourceProfile resourceProfile, int pendingSlots) {
	int currentPendingSlots = pendingSlots;
	while (currentPendingSlots < getNumberRequiredTaskManagerSlots()) {
		final Collection<ResourceProfile> slots = tryStartNewWorker(resourceProfile);

		if (slots.isEmpty()) {
			break;
		}

		currentPendingSlots += slots.size();
	}
}
```